includes/class-wc-gateway-stripe.php

```diff
@@ -1136,22 +1148,111 @@ public function get_saved_payment_method_option_html( $token ) {
 	 * before displaying the Thank You page. This is useful in case webhooks have not been set up.
 	 *
 	 * @since 4.2.0
-	 *
 	 * @param int $order_id The ID that will be used for the thank you page.
 	 * @return int          The ID to use.
 	 */
 	public function thankyou_order_id( $order_id ) {
 		if ( $order_id > 0 ) {
 			$order = wc_get_order( $order_id );
-
-			if ( $order && 'pending' === $order->get_status() && $order->get_payment_method() === $this->id ) {
-				$intent = $this->get_intent_from_order( $order );
-				if ( 'succeeded' === $intent->status ) {
-					$this->process_response( end( $intent->charges->data ), $order );
-				}
-			}
+			$this->verify_intent_after_checkout( $order );
 		}
 
 		return $order_id;
 	}
+
+	/**
+	 * Attached to `woocommerce_payment_successful_result` with a late priority,
+	 * this method will combine the "naturally" generated redirect URL from
+	 * WooCommerce and a payment intent secret into a hash, which contains both
+	 * the secret, and a proper URL, which will confirm whether the intent succeeded.
+	 *
+	 * @since 4.2.0
+	 * @param array $result   The result from `process_payment`.
+	 * @param int   $order_id The ID of the order which is being paid for.
+	 * @return array
+	 */
+	public function modify_successful_payment_result( $result, $order_id ) {
+		// Only redirects with intents need to be modified.
+		if ( ! isset( $result['intent_secret'] ) ) {
+			return $result;
+		}
+
+		// Put the final thank you page redirect into the verification URL.
+		$verification_url = add_query_arg(
+			array(
+				'order'       => $order_id,
+				'nonce'       => wp_create_nonce( 'wc_stripe_confirm_pi' ),
+				'redirect_to' => rawurlencode( $result['redirect'] ),
+			),
+			WC_AJAX::get_endpoint( 'wc_stripe_verify_intent' )
+		);
+
+		// Combine into a hash.
+		$redirect = sprintf( '#confirm-pi-%s:%s', $result['intent_secret'], rawurlencode( $verification_url ) );
+
+		return array(
+			'result'   => 'success',
+			'redirect' => $redirect,
+		);
+	}
+
+	/**
+	 * Executed between the "Checkout" and "Thank you" pages, this
+	 * method updates orders based on the status of associated PaymentIntents.
+	 *
+	 * @since 4.2.0
+	 * @param WC_Order $order The order which is in a transitional state.
+	 */
+	public function verify_intent_after_checkout( $order ) {
+		if ( 'pending' !== $order->get_status() && 'failed' !== $order->get_status() ) {
+			// If payment has already been completed, this function is redundant.
+			return;
+		}
+
+		if ( $order->get_payment_method() !== $this->id ) {
+			// If this is not the payment method, an intent would not be available.
+			return;
+		}
+
+		$intent = $this->get_intent_from_order( $order );
+		if ( ! $intent ) {
+			// No intent, redirect to the order received page for further actions.
+			return;
+		}
+
+		if ( 'succeeded' === $intent->status ) {
+			// Proceed with the payment completion.
+			$this->process_response( end( $intent->charges->data ), $order );
+		} else {
```

Comment:
Could the intent be still in a `processing` status at this point? Like, not succeded, but not failed yet? Is there any possibility that the SCA flow will complete "asynchronously" and so the site **has to** rely on Webhooks? If that's a possibility, I would reconsider this logic (marking the order immediately as `failed`).

Reply:
I am sure I discussed the reliability on webhooks with @thorsten-stripe at one point and we don't _have to_ rely on them. If that was the case, even though it would have been much more complicated for our merchants, it would've made this whole PR redundant.

Additionally if that was the case, during my testing I should've seen a failed and then completed order at one point, which I didn't.

Reply:
This holds true for card payments, however in the future we will unify non-card payments in the PaymentIntent API, meaning then the PI could be in processing state for an unknown amount of time.

Reply:
Thank you for updating us on this @thorsten-stripe.

To avoid marking pending orders as failed, in 97aa482 I changed the behavior. Now we will only mark the order as failed when the PI status is `requires_payment_method`.

Note: This PR/branch does not affect any alternative payment methods at the moment, but it is likely for this code to be reused when PIs are implemented for non-card payments, which is why I added the change.

Reply:
> To avoid marking pending orders as failed, in 97aa482 I changed the behavior. Now we will only mark the order as failed when the PI status is requires_payment_method.

I see that in the `failed_sca_auth` function, a similar check is made, which is now redundant:
https://github.com/woocommerce/woocommerce-gateway-stripe/blob/5796dad8d21d70210bfd9ad525b6be04ec4806ce/includes/class-wc-gateway-stripe.php#L1379

Reply:
Good point. I removed the second line. `requires_action` should also not appear at that point, so the line really did become completely redundant.